firebase-firestore/src/test/java/com/google/firebase/firestore/local/SQLiteLocalStoreTest.java

```diff
@@ -290,4 +307,60 @@ public void testIndexesServerTimestamps() {
     assertOverlayTypes(keyMap("coll/a", CountingQueryEngine.OverlayType.Set));
     assertQueryReturned("coll/a");
   }
+
+  @Test
+  public void testDeeplyNestedServerTimestamps() {
+    Timestamp timestamp = Timestamp.now();
+    Value initialServerTimestamp = ServerTimestamps.valueOf(timestamp, null);
+    Map<String, Value> fields =
+        new HashMap<String, Value>() {
+          {
+            put("timestamp", ServerTimestamps.valueOf(timestamp, initialServerTimestamp));
+          }
+        };
+    FieldPath path = FieldPath.fromSingleSegment("timestamp");
+    FieldMask mask =
+        FieldMask.fromSet(
+            new HashSet<FieldPath>() {
+              {
+                add(path);
+              }
+            });
+    FieldTransform fieldTransform =
+        new FieldTransform(path, ServerTimestampOperation.getInstance());
+    List<FieldTransform> fieldTransforms =
+        new ArrayList<FieldTransform>() {
+          {
+            add(fieldTransform);
+          }
+        };
+    AtomicReference<Throwable> error = new AtomicReference<>();
+    Thread thread =
```

Comment:
Maybe explain in comments as well.

Reply:
It doesn't _need_ to be multi-threaded. In fact, if we remove the `Thread` and just run the `for` loop synchronously, the test fails without the fix, and passes with the fix (on my machine).

The only reason we're creating a thread here is because the Java `Thread` class allows us to set a specific `stackSize` ([the last parameter is the `stackSize`](https://docs.oracle.com/javase/7/docs/api/java/lang/Thread.html#Thread(java.lang.ThreadGroup,%20java.lang.Runnable,%20java.lang.String,%20long))). Note that the `stackSize` parameter is approximate and is not guaranteed to work on all machines, but it's better than nothing.

>The stack size is the approximate number of bytes of address space that the virtual machine is to allocate for this thread's stack. The effect of the stackSize parameter, if any, is highly platform dependent.
On some platforms, specifying a higher value for the stackSize parameter may allow a thread to achieve greater recursion depth before throwing a [StackOverflowError](https://docs.oracle.com/javase/7/docs/api/java/lang/StackOverflowError.html). Similarly, specifying a lower value may allow a greater number of threads to exist concurrently without throwing an [OutOfMemoryError](https://docs.oracle.com/javase/7/docs/api/java/lang/OutOfMemoryError.html) (or other internal error). The details of the relationship between the value of the stackSize parameter and the maximum recursion depth and concurrency level are platform-dependent. On some platforms, the value of the stackSize parameter may have no effect whatsoever.

Also note that when porting this fix to iOS (C++), there's [no way to specify the stack size](https://stackoverflow.com/a/13871812) (other than compiler flags) and I'll just implement the test to run the for loop synchronously.

Reply:
Thanks.